File: scripts/fix-logger-calls.js
#!/usr/bin/env node
/**
 * Fix Pino logger call signatures to match proper format:
 * log.level(message: string, meta?: Record<string, any>)
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

console.log('🔍 Finding files with logger signature errors...\n');

// Get list of files with TypeScript errors
let errorOutput;
try {
  execSync('NODE_OPTIONS="--max-old-space-size=8192" npx tsc --noEmit 2>&1', {
    encoding: 'utf-8',
    env: { ...process.env, NODE_OPTIONS: '--max-old-space-size=8192' },
  });
  console.log('✅ No TypeScript errors found!');
  process.exit(0);
} catch (error) {
  errorOutput = error.stdout || error.stderr || '';
}

// Extract unique file paths
const errorLines = errorOutput.split('\n').filter((line) => line.startsWith('src/'));
const files = [...new Set(errorLines.map((line) => line.split('(')[0]))].filter(Boolean).sort();

if (files.length === 0) {
  console.log('✅ No files with errors found!');
  process.exit(0);
}

console.log(`📊 Found ${files.length} files with errors`);
console.log('🔧 Fixing logger signatures...\n');

let fixedCount = 0;

files.forEach((file, index) => {
  const filePath = path.join(process.cwd(), file);

  if (!fs.existsSync(filePath)) {
    return;
  }

  console.log(`[${index + 1}/${files.length}] Processing: ${file}`);

  let content = fs.readFileSync(filePath, 'utf-8');
  const original = content;

  // Fix Pattern 1: log.level('message', simpleVariable)
  // Where simpleVariable is a simple identifier (not already an object)
  content = content.replace(
    /log\.(error|warn|info|debug)\(([^,]+),\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*\)/g,
    (match, level, message, variable) => {
      // Skip if variable name suggests it's already metadata
      if (variable === 'metadata' || variable === 'meta' || variable === 'data') {
        return match;
      }
      return `log.${level}(${message}, { ${variable} })`;
    }
  );

  // Fix Pattern 2: log.level('message', 'string literal')
  // Combine the message and literal
  content = content.replace(
    /log\.(error|warn|info|debug)\((['"`])([^'"`]+)\2,\s*(['"`])([^'"`]+)\4\s*\)/g,
    (match, level, quote1, msg, quote2, literal) => {
      return `log.${level}(${quote1}${msg} ${literal}${quote1})`;
    }
  );

  // Fix Pattern 3: log.level('message', variable1, variable2, ...)
  // Multiple arguments that need to be wrapped
  content = content.replace(
    /log\.(error|warn|info|debug)\(([^,]+),\s+([^)]+)\)/g,
    (match, level, message, restArgs) => {
      // Check if restArgs already looks like an object literal
      const trimmedRest = restArgs.trim();

      // If it's already an object literal, skip
      if (trimmedRest.startsWith('{') && trimmedRest.endsWith('}')) {
        return match;
      }

      // If it contains only one simple variable and we already processed it, skip
      if (/^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(trimmedRest)) {
        // Already handled by Pattern 1
        return match;
      }

      // Split by commas (simple split - won't handle nested calls perfectly but good enough)
      const args = restArgs.split(',').map((a) => a.trim());

      // If only one arg and it's a string literal, skip (Pattern 2 handles it)
      if (args.length === 1 && /^['"`]/.test(args[0])) {
        return match;
      }

      // Build metadata object
      const metadataEntries = args.map((arg) => {
        // If it's a simple variable, use shorthand
        if (/^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(arg)) {
          return arg;
        }
        // For complex expressions, we'll need a key
        // Try to extract a meaningful name
        const varMatch = arg.match(/([a-zA-Z_$][a-zA-Z0-9_$]*)/);
        if (varMatch) {
          return `${varMatch[1]}: ${arg}`;
        }
        return `value: ${arg}`;
      });

      return `log.${level}(${message}, { ${metadataEntries.join(', ')} })`;
    }
  );

  if (content !== original) {
    fs.writeFileSync(filePath, content, 'utf-8');
    console.log('  ✅ Fixed\n');
    fixedCount++;
  } else {
    console.log('  ⏭️  No changes\n');
  }
});

console.log(`✅ Processing complete! Fixed ${fixedCount} files\n`);

// Re-run TypeScript check
console.log('⚠️  Running TypeScript check to verify fixes...\n');

try {
  execSync('NODE_OPTIONS="--max-old-space-size=8192" npx tsc --noEmit 2>&1', {
    encoding: 'utf-8',
    stdio: 'inherit',
    env: { ...process.env, NODE_OPTIONS: '--max-old-space-size=8192' },
  });
  console.log('\n✅ All logger signature errors fixed!');
  process.exit(0);
} catch (error) {
  const output = error.stdout || error.stderr || '';
  const remainingErrors = output.split('\n').filter((line) => line.startsWith('src/')).length;

  console.log(`\n📊 Remaining TypeScript errors: ${remainingErrors}`);

  if (remainingErrors > 0) {
    console.log('⚠️  Some errors remain - saving to /tmp/remaining-logger-errors.log');
    fs.writeFileSync('/tmp/remaining-logger-errors.log', output, 'utf-8');
    process.exit(1);
  }

  process.exit(0);
}

File: scripts/fix-logger-calls.py
#!/usr/bin/env python3
"""
Fix Pino logger call signatures to match proper format:
log.level(message: string, meta?: Record<string, any>)
"""

import re
import sys
from pathlib import Path

def fix_logger_call(match):
    """Fix a single logger call to proper format"""
    level = match.group(1)  # error, warn, info, debug
    args = match.group(2)   # everything between parentheses

    # Split arguments by comma (but not commas inside quotes or objects)
    arg_parts = []
    current_arg = ""
    paren_depth = 0
    brace_depth = 0
    in_string = False
    string_char = None

    for char in args:
        if char in ('"', "'", '`') and not in_string:
            in_string = True
            string_char = char
        elif char == string_char and in_string:
            in_string = False
            string_char = None
        elif char == '(' and not in_string:
            paren_depth += 1
        elif char == ')' and not in_string:
            paren_depth -= 1
        elif char == '{' and not in_string:
            brace_depth += 1
        elif char == '}' and not in_string:
            brace_depth -= 1
        elif char == ',' and not in_string and paren_depth == 0 and brace_depth == 0:
            arg_parts.append(current_arg.strip())
            current_arg = ""
            continue

        current_arg += char

    if current_arg.strip():
        arg_parts.append(current_arg.strip())

    # If already correct format (1 or 2 args, second is object), return as-is
    if len(arg_parts) <= 1:
        return match.group(0)

    if len(arg_parts) == 2:
        second_arg = arg_parts[1].strip()
        # Check if second arg is already an object literal
        if second_arg.startswith('{') and second_arg.endswith('}'):
            return match.group(0)

    # Fix needed: wrap additional arguments in metadata object
    message = arg_parts[0]
    rest_args = arg_parts[1:]

    # Create metadata object entries
    metadata_entries = []
    for i, arg in enumerate(rest_args):
        arg = arg.strip()
        # If it's a simple variable name, use it as shorthand property
        if re.match(r'^[a-zA-Z_][a-zA-Z0-9_]*$', arg):
            metadata_entries.append(arg)
        else:
            # For complex expressions, create numbered keys
            metadata_entries.append(f'arg{i + 1}: {arg}')

    # Build fixed call
    metadata_obj = '{ ' + ', '.join(metadata_entries) + ' }'
    fixed_call = f'log.{level}({message}, {metadata_obj})'

    return fixed_call

def fix_file(file_path):
    """Fix all logger calls in a file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        original_content = content

        # Pattern to match log.level(...) calls
        # Matches: log.error(...), log.warn(...), log.info(...), log.debug(...)
        pattern = r'log\.(error|warn|info|debug)\(([^)]+(?:\([^)]*\))?[^)]*)\)'

        # Replace all matches
        content = re.sub(pattern, fix_logger_call, content)

        if content != original_content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            return True
        return False

    except Exception as e:
        print(f'  ❌ Error processing {file_path}: {e}')
        return False

def main():
    # Get list of files with TypeScript errors
    import subprocess

    print('🔍 Finding files with logger signature errors...')

    try:
        result = subprocess.run(
            ['npx', 'tsc', '--noEmit'],
            capture_output=True,
            text=True,
            env={'NODE_OPTIONS': '--max-old-space-size=8192'}
        )

        # Extract unique file paths from errors
        error_lines = [line for line in result.stderr.split('\n') if line.startswith('src/')]
        files = set()
        for line in error_lines:
            if '(' in line:
                file_path = line.split('(')[0]
                files.add(file_path)

        files = sorted(files)

        if not files:
            print('✅ No TypeScript errors found!')
            return 0

        print(f'📊 Found {len(files)} files with errors')
        print('🔧 Fixing logger signatures...\n')

        fixed_count = 0
        for i, file_path in enumerate(files, 1):
            if not Path(file_path).exists():
                continue

            print(f'[{i}/{len(files)}] Processing: {file_path}')
            if fix_file(file_path):
                fixed_count += 1
                print(f'  ✅ Fixed')
            else:
                print(f'  ⏭️  No changes')

        print(f'\n✅ Processing complete! Fixed {fixed_count} files')

        # Re-run TypeScript check to see remaining errors
        print('\n⚠️  Running TypeScript check to verify fixes...\n')
        result = subprocess.run(
            ['npx', 'tsc', '--noEmit'],
            capture_output=True,
            text=True,
            env={'NODE_OPTIONS': '--max-old-space-size=8192'}
        )

        remaining_errors = len([line for line in result.stderr.split('\n') if line.startswith('src/')])
        print(f'\n📊 Remaining TypeScript errors: {remaining_errors}')

        if remaining_errors == 0:
            print('✅ All logger signature errors fixed!')
            return 0
        else:
            print('⚠️  Some errors remain - saving to /tmp/remaining-logger-errors.log')
            with open('/tmp/remaining-logger-errors.log', 'w') as f:
                f.write(result.stderr)
            return 1

    except Exception as e:
        print(f'❌ Error: {e}')
        return 1

if __name__ == '__main__':
    sys.exit(main())
